torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -2537,6 +2567,57 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
   }
 }
 
+void ProcessGroupNCCL::allgatherUniqueNCCLID(
+    int rootRank,
+    ncclUniqueId* ncclID,
+    std::vector<ncclUniqueId>& ncclIDs) {
+  std::vector<std::string> storeKeys;
+  for (size_t r = 0; r < ncclIDs.size(); r++) {
+    storeKeys.emplace_back("UniqueNCCLID:" + std::to_string(r));
+  }
+  if (rank_ == rootRank) {
+    auto vec = std::vector<uint8_t>(
+        reinterpret_cast<uint8_t*>(ncclID),
+        reinterpret_cast<uint8_t*>(ncclID) + NCCL_UNIQUE_ID_BYTES);
+    store_->set(storeKeys[rootRank], vec);
+  }
+  try {
+    int indx = 0;
+    for (auto& value : store_->multiGet(storeKeys)) {
+      TORCH_CHECK_WITH(
+          DistBackendError,
+          value.size() == NCCL_UNIQUE_ID_BYTES,
+          "Invalid size for ncclUniqueId");
+      std::memcpy(&ncclIDs[indx++], value.data(), value.size());
+    }
+  } catch (const std::exception& e) {
+    nlohmann::json json_vec = storeKeys;
+    std::string exceptionMsg = c10::str(
+        "[",
+        rank_,
+        "] is setting up NCCL communicators and "
+        "retrieving ncclUniqueId from roots via TCPStore by key '",
+        json_vec.dump(),
+        "', but got error: ");
+    C10_THROW_ERROR(
+        DistBackendError,
+        exceptionMsg + e.what() +
+            ". This may indicate a possible application crash on rank 0 or a network set up issue.");
+  } catch (...) {
+    nlohmann::json json_vec = storeKeys;
+    C10_THROW_ERROR(
+        DistBackendError,
+        c10::str(
+            "Unknown exception while [",
+            rank_,
+            "] is setting up NCCL communicators and "
+            "retrieving ncclUniqueIds from roots via TCPStore by key '",
+            json_vec.dump(),
+            "'",
+            ". This may indicate a possible application crash on rank 0 or a network set up issue."));
+  }
```

Comment:
nit: I wonder if there is a way to avoid duplicated code. (It may be hard, and I haven't done any research either)

Reply:
I thought about it. Unfortunately, since the logging message and API usage is different. I decide to not to reuse the existing code.

Reply:
Sorry about the confusion. I meant dedup the two catch blocks here.

Reply:
I don't think so, this is literally to generate logs and throw exceptions..